include/seqan3/index/fm_index_iterator.hpp

```diff
@@ -0,0 +1,521 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Christopher Pockrandt <christopher.pockrandt AT fu-berlin.de>
+ * \brief Contains the seqan3::fm_index_iterator used to search in the unidirectional seqan3::fm_index.
+ */
+
+#pragma once
+
+#include <array>
+
+#include <sdsl/suffix_trees.hpp>
+
+#include <range/v3/view/iota.hpp>
+#include <range/v3/view/slice.hpp>
+#include <range/v3/view/transform.hpp>
+
+#include <seqan3/alphabet/all.hpp>
+#include <seqan3/index/detail/fm_index_iterator.hpp>
+#include <seqan3/index/concept.hpp>
+#include <seqan3/std/concept/range.hpp>
+#include <seqan3/std/concept/iterator.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+
+namespace seqan3
+{
+
+/*!\addtogroup index
+ * \{
+ */
+
+// TODO: unify terminology (pattern, sequence)
+// TODO: remove mapping by overwriting backward_search.
+// one only has to deal with incomplete alphabets then (maybe add own alphabet type to sdsl?)
+// TODO: to_rank() + 1 consistent with comp_char, mapping and implicit_sentinel?
+
+/*!\brief The undirectional FM index iterator.
+ * \ingroup fm_index
+ * \implements seqan3::fm_index_iterator_concept
+ * \tparam index_t The type of the underlying index; must satisfy seqan3::fm_index_concept.
+ * \details
+ *
+ * The iterator's interface and behaviour is equivalent to a suffix tree with the space and time efficiency of the
+ * underlying pure FM index. The implicit suffix tree is not compacted, i.e. going down an edge will increase the
+ * searched sequence by only one character.
+ *
+ * The iterator traverses the implicit suffix tree beginning at the root node. All methods
+ * modifying the iterator (e.g. going down an edge) return a `bool` value whether the operation was successful or not.
+ * In case of an unsuccessful operation the iterator remains unmodified, i.e. there an iterator can never be in an
+ * invalid state except default constructed iterators that are always invalid. E.g. going down an edge while the
+ * iterator is in a leaf node, will return `false` and will remain in that leaf node.
+ *
+ * The asymptotic running times for using the iterator depend on the SDSL index configuration. To determine the exact
+ * running times, you have to look up the running times of the used traits (configuration) additionally.
```

Comment:
you have to **additionally** look up the running times of the used traits (configuration) ~~additionally~~.